### DISCUSSION TITLE
Average Colour of Specified Regions of Image

### DISCUSSION THREAD

**Discussion by uriel-tannen** [ImageMagick/ImageMagick#3257] [Help]
I'm trying to find the average RGB colour of specified regions of a set of images. I've got the following code which outputs the average colour of an entire image, but this clearly doesn't serve to output the average colour of specified regions:

```
(for filename in *.jpg; do 
    magick $filename -set colorspace sRGB -format "%f,%[fx:int(mean.r*255)],%[fx:int(mean.g*255)],%[fx:int(mean.b*255)]\n" info: >> "swatchesRGB.csv"
done) 
```

So for example the below image is 1088x1088, and I'd like to get the average colour of the following 3 regions:

- a 4x4px polygon that starts at [24,16]
- a 2x26px polygon that starts at [45,45]
- a 5x3px polygon that starts at [50,60]

I want to output different regions at different [x,y] co-ordinates for each image in my set, so I need to be able to feed in the relevant polygon sizes and co-ordinates for each image being processed.

I'm at a loss on how to implement this...any help would be much appreciated!

FYI my IM version is 7.0.10-45 Q16 x54 2020-11-30

![image](https://user-images.githubusercontent.com/36728824/108066023-b4568700-7056-11eb-97b9-80ef3731f253.png)

**Comment by snibgo** [comment]
The obvious way is to crop the image, and find statistics for that.

**Reply by uriel-tannen** [reply]
Thanks @snibgo - fair point. Given that each of the images will have multiple and different crop-sizes and starting co-ordinates, is there a way to feed into IM the relevant inputs from, say, a CSV file? 

In my ideal setup, I'd have a CSV file structured as follows:

![image](https://user-images.githubusercontent.com/36728824/108103177-7a4eaa80-7081-11eb-95f4-3c7578cffacc.png)

It would be nice to have a single script that crops each of the images listed in the first column using the specifications in the subsequent columns...is that possible?

**Comment by snibgo** [comment]
I would use a shell script that reads the CSV file. For each input image, it does one of these:

1. Builds an IM command in an environment variable, then executes that variable.

2. Builds a shell script that runs an IM command, then runs that script.

3. Builds an IM script, then runs "magick" with "-script".

The IM comand could be something like this (Windows BAT syntax):

```
%IMG7%magick ^
  chair.png ^
  ( -clone 0 -crop 200x20+545+213 +repage ) ^
  ( -clone 0 -crop 45x45+158+97 +repage ) ^
  -delete 0 ^
  -format "%%f,%%[fx:int(mean.r*255)],%%[fx:int(mean.g*255)],%%[fx:int(mean.b*255)]\n" ^
  info:
```
The output is something like:
```
chair.png,255,255,254
chair.png,255,255,255
```

Beware: if the cropped area includes transparency, the result will be misleading, as it ignores the alpha channel. You can "-scale" to find the means of _just_ the opaque pixels:
```
magick ^
  chair.png ^
  ( -clone 0 -crop 200x20+545+213 +repage -scale "1x1^!" ) ^
  ( -clone 0 -crop 45x45+158+97 +repage -scale "1x1^!" ) ^
  -delete 0 ^
  -format "%%f,%%[fx:int(mean.r*255)],%%[fx:int(mean.g*255)],%%[fx:int(mean.b*255)]\n" ^
  info:
```

**Reply by uriel-tannen** [reply]
Excellent @snibgo - this has really helped! Thank you.
